### PR TITLE
Allow partial credit for words which are right, apart from having the wrong accents

### PR DESCRIPTION
**Pull request by BruceGoodGuy**
Hello @timhunt ,

This pull request contains two commits:

**Commit 1** (ad6a724db3d19627f2f93b579b4cb52ad38dfcef): Implements a new feature - Allow the use of the same word with or without special characters and be able to adjust marks.

To resolve this, I added two new columns accentedlettersoptions and penaltyforincorrectaccents:
- accentedlettersoptions (int): Store accented letters options (0 => The answer must be completely correct and must not be accents wrong, 1 => Accents errors are allowed, but points will be deducted and 2 => Accents errors are allowed and the points will not be deducted).
- penaltyforincorrectaccents: store grade for answers with incorrect accents (from 5% to 90%).

newUI1

So the new UI will look like: 
newUI
newUI2

To detect the input answer is wrong accented only. I have written a new function called remove_accent (classes/util.php) This function just does one thing, which is to remove every accented character from the string input. The principle of the algorithm detecting the answer is wrong accented only is: We will use function remove_accent to remove every accented character of the right answer and user input answer then compare it together.

Then, I updated function get_num_parts_right (question.php). I updated a new way to calculate the number of right answers ($numright) and the number of partial answers ($numpartial) (answer only wrong accented characters) After; I updated grade_response (question.php) to adapt with the new function get_num_parts_right;

Then, I created a new function calculate_fraction_for_answer (renderer.php) to calculate the fraction per each answer.

Finally, I created and updated Unittest and Behat test for the new update to make sure every function works well.

**Commit 2:** (5553421dcab9505e65eba60b886e10ea492a26a1) Because of the request of Git action, I created a new commit to update postgres version and coding style.

Could you please help me review this pull request? If you have any concerns, please let me know. Thanks.

### CODE REVIEW

File: classes/util.php [hunk 1]
```diff
@@ -48,5 +61,42 @@ public static function safe_normalize(string $string): string {
...
-}
 
+    /**
+     * Remove accent character in text. Eg: Français -> Francais.
+     *
+     * @param string $string The input string.
+     * @return string The normal string without accents.
+     */
+    public static function remove_accent(string $string): string {
+        return preg_replace('/\p{Mn}/u', '', self::safe_normalize($string, Normalizer::FORM_KD));
```
Comment (timhunt) [inline classes/util.php:72]: Is it true that every Mn = Mark, nonspacing unicode character is an accent?

**Review by timhunt — CHANGES_REQUESTED** [review]
I am not claiming this is a complete review, but having found more than 10 things to comment on, I have had enough. I will look again once these things are fixed, plus anything else you find yourself when you review this.

File: classes/util.php [hunk 2]
```diff
@@ -39,7 +52,7 @@ public static function safe_normalize(string $string): string {
             return $string;
```
[the remark this answers is not in the document]
Comment (BruceGoodGuy) [reply classes/util.php]: I have removed it.

File: backup/moodle2/backup_qtype_crossword_plugin.class.php [hunk 3]
```diff
@@ -39,7 +39,7 @@ protected function define_question_plugin_structure(): backup_plugin_element {
 
         // Now create the qtype own structures.
         $crossword = new backup_nested_element('crossword', ['id'], ['correctfeedback',
-            'correctfeedbackformat', 'numrows', 'numcolumns',
+            'correctfeedbackformat', 'numrows', 'numcolumns', 'accentedlettersoptions', 'penaltyforincorrectaccents',
```
[the remark this answers is not in the document]
Comment (BruceGoodGuy) [reply backup/moodle2/backup_qtype_crossword_plugin.class.php]: I have changed it to accentgradingtype  and accentpenalty.

File: classes/util.php [hunk 4]
```diff
@@ -24,13 +26,24 @@
...
 class util {
+
+    /** @var int The answer must be completely correct and must not be accents wrong */
+    const DONT_ACCEPT_WRONG_ACCENTED = 0;
+
+    /** @var int Accents errors are allowed, but points will be deducted. */
+    const ACCEPT_WRONG_ACCENTED_BUT_PENALTY = 1;
+
+    /** @var int Accents errors are allowed and the points will not be deducted. */
+    const ACCEPT_WRONG_ACCENTED = 2;
```
[the remark this answers is not in the document]
Comment (BruceGoodGuy) [reply classes/util.php]: - I have moved these constants to question/type/crossword/questiontype.php.
- I have changed the name of constants and their value. So the new constants will look likes:
image

File: lang/en/qtype_crossword.php [hunk 5]
```diff
@@ -22,21 +22,27 @@
  * @license  https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
 
+$string['accentedcharacters'] = 'Accented letters';
+$string['acceptwrongaccents'] = 'Just grade the letters and ignore any accents';
 $string['across'] = 'Across';
```
[the remark this answers is not in the document]
Comment (BruceGoodGuy) [reply lang/en/qtype_crossword.php:30]: I have moved it into group:
image

File: question.php [hunk 6]
```diff
@@ -108,19 +116,34 @@ public function is_same_response(array $prevresponse, array $newresponse): bool
     }
 
     public function grade_response(array $response): array {
-        list($right, $total) = $this->get_num_parts_right($response);
-        $fraction = $right / $total;
+        // Retrieve a number of right answers and only wrong accent answers.
+        [$numrightanswers, $numpartialanswers] = $this->get_num_parts_right($response);
```
[the remark this answers is not in the document]
Comment (BruceGoodGuy) [reply question.php]: I adjusted the get_num_parts_right function, now it only returns the number of correct answers and the total number of answers.

File: classes/util.php [hunk 7]
```diff
@@ -39,7 +52,7 @@ public static function safe_normalize(string $string): string {
             return $string;
         }
 
-        $normalised = normalizer_normalize($string, Normalizer::FORM_KC);
+        $normalised = normalizer_normalize($string, $normalizeform);
         if (is_null($normalised)) {
```
[the remark this answers is not in the document]
Comment (BruceGoodGuy) [reply classes/util.php]: I changed the condition logic for this case:
image

File: classes/util.php [hunk 8]
```diff
@@ -48,5 +61,42 @@ public static function safe_normalize(string $string): string {
...
+
+    /**
+     * Calculate fraction of answer.
+     *
+     * @param qtype_crossword_question $question The question object.
+     * @param answer $answer The answer object.
+     * @param string $inputanswer The inputanswer need to calculate.
+     * @return float The fraction value of the answer.
+     */
+    public static function calculate_fraction_for_answer(qtype_crossword_question $question,
```
[the remark this answers is not in the document]
Comment (BruceGoodGuy) [reply classes/util.php]: I've adjust the code to make it easier to read and also prevent duplicate code.

File: version.php [hunk 9]
```diff
@@ -25,9 +25,9 @@
 defined('MOODLE_INTERNAL') || die();
 
 $plugin->component = 'qtype_crossword';
-$plugin->version   = 2022101000;
```
[the remark this answers is not in the document]
Comment (BruceGoodGuy) [reply version.php:28]: I have reverted the code changes.

File: tests/util_test.php [hunk 10]
```diff
@@ -72,4 +72,175 @@ public function test_safe_normalize_provider(): array {
...
         ];
     }
+
+    /**
+     * Test remove_accent function.
+     *
+     * @dataProvider test_remove_accent_provider
+     * @covers \qtype_crossword\util::remove_accent
+     *
+     * @param string $string1 The first string need to compare.
```
[the remark this answers is not in the document]
Comment (BruceGoodGuy) [reply tests/util_test.php]: I'm sorry, I have updated it.

File: tests/question_test.php [hunk 11]
```diff
@@ -108,16 +110,19 @@ public function test_is_gradable_response() {
...
-        $this->assertEquals([1, \question_state::$gradedright],
-            $question->grade_response(['sub0' => 'BRAZIL', 'sub1' => 'PARIS', 'sub2' => 'ITALY']));
-
+    public function test_grading(array $answeroptions) {
+        $question = \test_question_maker::make_question('crossword', 'not_accept_wrong_accents');
+        $question->accentedlettersoptions = $answeroptions['options']['accentedlettersoptions'];
+        $question->penaltyforincorrectaccents = $answeroptions['options']['penalty'];
+        foreach ($answeroptions['answers'] as $answer) {
+            [$fraction, $state] = $question->grade_response($answer['answers']);
+            $this->assertEquals($answer['fraction'], $fraction);
```
[the remark this answers is not in the document]
Comment (BruceGoodGuy) [reply tests/question_test.php]: Thanks, I have updated it.
image

File: classes/util.php (hunk 1 above)
Comment (BruceGoodGuy) [reply classes/util.php:58]: This function will remove all accents, not only for Europe but also other regions like Asian accents,... and even [TONE MARKS](https://vietnamesetypography.com/tone-marks/). Should I adjust this function to remove only Europe accent character?

**Comment by BruceGoodGuy** [comment]
Updated:
I rebased this pull request with the latest code. 
Besides that, I removed `tests/behat/backup_and_restore.feature` file, backup/restore feature will be covered in the unit test `tests/backup_test.php`.

And at line 61 (backup/moodle2/backup_qtype_crossword_plugin.class.php), to fix an issue can not load images after backup/restore. Because the code change is too small, I merged it into this commit.

Could you please help me to review it? thanks @timhunt .

**Comment by timhunt** [comment]
I merged this manually, so closing.

**Pull request closed without merging**